src/basic/proc-cmdline.c

```diff
@@ -303,75 +303,91 @@ int proc_cmdline_get_bool(const char *key, bool *ret) {
         return 1;
 }
 
-int proc_cmdline_get_key_many_internal(ProcCmdlineFlags flags, ...) {
-        _cleanup_free_ char *line = NULL;
-        bool processing_efi = true;
-        const char *p;
-        va_list ap;
-        int r, ret = 0;
+typedef struct ProcCmdlineDataEntry {
+        const char *key;
+        char *value;
+        char **ret;
+} ProcCmdlineDataEntry;
 
-        /* The PROC_CMDLINE_VALUE_OPTIONAL flag doesn't really make sense for proc_cmdline_get_key_many(), let's make
-         * this clear. */
-        assert(!FLAGS_SET(flags, PROC_CMDLINE_VALUE_OPTIONAL));
+typedef struct ProcCmdlineData {
+        ProcCmdlineDataEntry *data;
+        size_t num;
+} ProcCmdlineData;
 
-        /* This call may clobber arguments on failure! */
+static void proc_cmdline_data_done(ProcCmdlineData *c) {
+        assert(c);
 
-        if (!FLAGS_SET(flags, PROC_CMDLINE_IGNORE_EFI_OPTIONS)) {
-                r = systemd_efi_options_variable(&line);
-                if (r < 0 && r != -ENODATA)
-                        log_debug_errno(r, "Failed to get SystemdOptions EFI variable, ignoring: %m");
-        }
+        FOREACH_ARRAY(e, c->data, c->num)
+                free(e->value);
 
-        p = line;
-        for (;;) {
-                _cleanup_free_ char *word = NULL;
+        free(c->data);
+}
 
-                r = proc_cmdline_extract_first(&p, &word, flags);
-                if (r < 0)
-                        return r;
-                if (r == 0) {
-                        /* We finished with this command line. If this was the EFI one, then let's proceed with the regular one */
-                        if (processing_efi) {
-                                processing_efi = false;
+static int get_many_handler(const char *key, const char *value, void *data) {
+        ProcCmdlineData *c = ASSERT_PTR(data);
+        int r;
 
-                                line = mfree(line);
-                                r = proc_cmdline(&line);
-                                if (r < 0)
-                                        return r;
+        assert(key);
 
-                                p = line;
-                                continue;
-                        }
+        FOREACH_ARRAY(e, c->data, c->num) {
+                if (!proc_cmdline_key_streq(e->key, key))
+                        continue;
 
-                        break;
+                if (value) {
+                        r = free_and_strdup(&e->value, value);
```

Comment:
Hmm, I would expect us to replace `e->value` unconditionally. I.e. drop the `if`.